src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupLinearFunctional.java

```diff
@@ -0,0 +1,516 @@
+				final int[] aix = sb.indexes(row);
+				final double[] avals = sb.values(row);
+				final int offRes = lhs._colIndexes[row] * result.getNumColumns();
+				for(int col = apos; col < alen; col++)
+					resV[offRes + _colIndexes[aix[col]]] += avals[col];
+			}
+		}
+		else {
+			double[] tmpRetV = tmpRet.getDenseBlockValues();
+			for(int row = 0; row < lhs.getNumCols(); row++) {
+				final int offRes = lhs._colIndexes[row] * result.getNumColumns();
+				final int offTmp = row * getNumCols();
+				for(int col = 0; col < getNumCols(); col++) {
+					resV[offRes + _colIndexes[col]] += tmpRetV[offTmp + col];
+				}
+			}
+		}
+	}
+
+	@Override
+	public void tsmmAColGroup(AColGroup other, MatrixBlock result) {
+		throw new DMLCompressionException("Should not be called");
+	}
+
+	@Override
+	protected AColGroup sliceSingleColumn(int idx) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	protected AColGroup sliceMultiColumns(int idStart, int idEnd, int[] outputCols) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public AColGroup copy() {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public boolean containsValue(double pattern) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public long getNumberNonZeros(int nRows) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public AColGroup replace(double pattern, double replace) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public void readFields(DataInput in) throws IOException {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public void write(DataOutput out) throws IOException {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public long getExactSizeOnDisk() {
+		long ret = super.getExactSizeOnDisk();
+		ret += _coefficents[0].length * _coefficents.length * 8L;
+		ret += 4L; // _numRows
+		return ret;
+	}
+
+	@Override
+	protected void computeProduct(double[] c, int nRows) {
+		c[0] = 1;
+		for(int colIndex : _colIndexes) {
+			double intercept = _coefficents[0][colIndex];
+			double slope = _coefficents[1][colIndex];
+			for(int i = 0; i < nRows; i++) {
+				c[0] *= intercept + slope * (i + 1);
+			}
+		}
+	}
+
+	@Override
+	protected void computeRowProduct(double[] c, int rl, int ru, double[] preAgg) {
+		for(int rix = rl; rix < ru; rix++) {
+			c[rix] = 1;
+			for(int colIndex : _colIndexes) {
+				double intercept = _coefficents[0][colIndex];
+				double slope = _coefficents[1][colIndex];
+				c[rix] *= intercept + slope * (rix + 1);
+			}
+		}
+	}
+
+	@Override
+	protected void computeColProduct(double[] c, int nRows) {
+		for(int colIndex : _colIndexes) {
+			c[colIndex] = 1;
+			double intercept = _coefficents[0][colIndex];
+			double slope = _coefficents[1][colIndex];
+			for(int i = 0; i < nRows; i++) {
+				c[colIndex] *= intercept + slope * (i + 1);
+			}
+		}
+	}
+
+	@Override
+	protected double[] preAggSumRows() {
+		throw new NotImplementedException();
```

Comment:
use these methods to allocate what you want to use in the SumRows call.

for instance you currently have:

```
	@Override
	protected void computeRowSums(double[] c, int rl, int ru, double[] preAgg) {
	        double intercept_sum = 0;
		for(double intercept : _coefficents[0])
			intercept_sum += intercept;

		double slope_sum = 0;
		for(double slope : _coefficents[1])
			slope_sum += slope;

		for(int rix = rl; rix < ru; rix++)
			c[rix] += intercept_sum + slope_sum * (rix + 1);
        }
```

instead of processing all the coefficients in all calls to all rl to ru,
preAggregate the first two for loops into the double[] array to contain two values `double[]{intercept_sum, slope_sum}`
this makes the parallelization of different row ranges trivial, and much much much faster.